iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,274 @@
+namespace iox
+{
+template <uint64_t Capacity, typename ValueType>
+IndexQueue<Capacity, ValueType>::IndexQueue(ConstructEmpty_t)
+    : m_readPosition(Index(Capacity))
+    , m_writePosition(Index(Capacity))
+{
+}
+
+template <uint64_t Capacity, typename ValueType>
+IndexQueue<Capacity, ValueType>::IndexQueue(ConstructFull_t)
+    : m_readPosition(Index(0))
+    , m_writePosition(Index(Capacity))
+{
+    for (uint64_t i = 0; i < Capacity; ++i)
+    {
+        m_cells[i].store(Index(i));
+    }
+}
+
+template <uint64_t Capacity, typename ValueType>
+constexpr uint64_t IndexQueue<Capacity, ValueType>::capacity()
+{
+    return Capacity;
+}
+
+template <uint64_t Capacity, typename ValueType>
+void IndexQueue<Capacity, ValueType>::push(ValueType index)
+{
+    // we need the CAS loop here since we may fail due to concurrent push operations
```

Comment:
You are writing here is somewhere a CAS loop hidden?! I have to check every method call inside this method to get the memory order and find all the atomic access and this really irritates me and makes these lock free code even harder to understand, to fix and when we do mistakes with this approach they are even harder to spot.
Please do not do this.

My wish would be that every relevant atomic operation is right in place in this function

Reply:
The CAS loop is the main and only loop of the method. The comment shall indicate that it is needed because the operation can fail due to concurrent operations. The code abstraction is very much a matter of opinion. It basically was the way you mention a long long time ago...when dinosaurs walked the earth (in march this year if i recall correctly). We can argue both ways, the way it is, the messy details are hidden and (maybe) the intention of the call captured by its name (the call overhead will most likely be optimized away). On the other hand it hides important algorithmic details inside those helpers. So, Clean Code ... quo vadis?